### ISSUE TITLE
Allow building prost-build without requiring curl & OpenSSL

### ISSUE THREAD

**Issue by briansmith** [danburkert/prost#77]
For a variety of reasons, including improving the build time of my project, I would like to build prost-build without downloading and building curl and its dependencies, particularly OpenSSL. In particular, I would like to avoid the build failure I get:

```text
error: failed to run custom build command for `openssl-sys v0.9.26`
process didn't exit successfully: `/home/m/go/src/github.com/runconduit/conduit/target/debug/build/openssl-sys-6bc9b54748dee815/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_LIB_DIR
cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_INCLUDE_DIR
cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
cargo:rerun-if-env-changed=X86_64_UNKNOWN_LINUX_GNU_OPENSSL_DIR
cargo:rerun-if-env-changed=OPENSSL_DIR
run pkg_config fail: "Failed to run `\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"`: No such file or directory (os error 2)"

--- stderr
thread 'main' panicked at '

Could not find directory of OpenSSL installation, and this `-sys` crate cannot
proceed without this knowledge. If OpenSSL is installed and this crate had
trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
compilation process.

If you're in a situation where you think the directory *should* be found
automatically, please open a bug at https://github.com/sfackler/rust-openssl
and include information about your system as well as this message.

    $HOST = x86_64-unknown-linux-gnu
    $TARGET = x86_64-unknown-linux-gnu
    openssl-sys = 0.9.26


It looks like you're compiling on Linux and also targeting Linux. Currently this
requires the `pkg-config` utility to find OpenSSL but unfortunately `pkg-config`
could not be found. If you have OpenSSL installed you can likely fix this by
installing `pkg-config`.

', /home/m/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-sys-0.9.26/build.rs:213:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

warning: build failed, waiting for other jobs to finish...
error: build failed
```

I don't want to require users to install pkg_config or do *any* manual configuration.

In fact, we already require protoc for the Go code in our project, so it would be ideal for us if we could download protoc ourselves and build prost-build without *any* downloading logic included at all.

If it's not acceptable to provide a way to completely avoid the downloading logic then maybe #27 might help? However, since our project is security-sensitive it would also be important to verify the SHA-256 checksums of the downloaded files, instead of just relying on HTTPS, so I think additional work would be necessary.

**Comment by briansmith** [comment]
Yes, but that doesn't prevent building the curl crate or the openssl crate. I want to prevent building those crates.

I also noticed:

* Compiling the curl-sys crate takes quite in our build, which is already painfully slow
* When I build openssl-1.0.2n from source and use it (passing the various OPENSSL_DIR environment variables), building prost-build fails with:

```
 /usr/bin/ld: /home/m/go/src/github.com/runconduit/conduit/target/debug/deps/libopenssl_sys-744e020eb234fbe1.rlib(ssl_lib.o): relocation R_X86_64_32 against `.rodata.str1.1' can not be used when making a shared object; recompile with -fPIC
          /home/m/go/src/github.com/runconduit/conduit/target/debug/deps/libopenssl_sys-744e020eb234fbe1.rlib: error adding symbols: Bad value
          collect2: error: ld returned 1 exit status
```

If I try to use the system packages then the build fails because I don't have the "dev" package for OpenSSL available, only the binary shared objects.

Basically, the curl dependency is very heavyweight and I'd really like to find a way to avoid it.

**Comment by danburkert** [comment]
Ah, I see.  So I'm on-board with switching over to Reqwest if that helps with build times, my only feedback on #27 is that it doesn't replace every use of curl in the codebase.  That being said I'm a bit confused because I think Reqwest still depends on openssl through hyper-tls, so is that going to fix the underlying issue?

Another option is to add curl as an optional (enabled by default) dependency of prost-build and re-structure the build script such that it isn't necessary if the environment variables are set.

**Comment by briansmith** [comment]
curl is a build dependency. I don't think it's possible to have a feature control a build dependency, is it?

`cargo fetch --help` says:

> If a lockfile is available, this command will ensure that all of the git
dependencies and/or registries dependencies are downloaded and locally
available. The network is never touched after a `cargo fetch` unless
the lockfile changes.

That is, we're supposed to be able to do `cargo fetch` and then disconnect from the internet and have a subsequent build succeed, or do the `cargo fetch` on a network-connected machine and the do the build of the downloaded things on a non-network-connected machine. In that spirit, I think the best option is to separate out the downloading of protoc completely from prost-build. This isn't possible if/when build scripts download code. This repo could provide a shell script (and/or batch file) that contains the same curl logic, that people could put into their build scripts. Obviously that would be a breaking change.

Alternatively, we could include all the protoc binaries in the protoc-build package. I do a similar thing in *ring*, where I include generated (binary) files in the crates.io package that aren't in the source code, though it's admittedly pretty ugly.

I'm not sure how to do it otherwise.

**Comment by danburkert** [comment]
Adding the precompiled `protoc` executables to the crate is an interesting option, I've never considered that.  It would only add about ~10MiB to the crate size, which I think is reasonable.  Do you have any pointers on how you can bundle arbitrary payloads into a crate without them being in the git tree?

**Comment by danburkert** [comment]
or maybe they should just be in-tree, since if they aren't you won't be able to build prost from source as easily?

**Comment by briansmith** [comment]
If you are OK with them being in the tree, that would be be the simplest option.

It turns out that I misremembered: Feature flags *can* control build dependencies. So, I'm working on a PR that simply makes the downloading optional based on a feature flag "download_protoc" since that's the least intrusive option.

**Comment by briansmith** [comment]
> I think eventually I want to explore shipping the bins as part of the crate, but that can come later.

Implementing the download as an optional feature isn't as trivial as I'd like. In particular, presumably if you add a feature you would want it to be included in the test matrix, and that gets messy quickly. Also, the feature has to be propogated throughout the project on anything that depends on "../prost-build", which also gets messy. I actually got most of the way to making the code (I think) but then I saw protobuf/ has its own downloading logic, and I haven't even started expanding the test matrix to include this option.

Given that, and given that we seem to both want "cargo fetch" semantics anyway, if you're OK vendoring the binaries into the GitHub repo, I'd rather just skip to that.

Would you be OK with just vendoring the amd64 binaries for Windows, macOS, and Linux for now, i.e. the contents of these files?:

* https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-win32.zip (I guess these are actually x86, not x86-64).
* https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-osx-x86_64.zip
* https://github.com/google/protobuf/releases/download/v3.5.1/protoc-3.5.1-linux-x86_64.zip

Or you want to have all platforms for which Google ships binary releases? Or something in between?

I notice that in #62 the dependency on protobuf was set to "3.5.1.1" in protobuf/ burt "3.5.1" elsewhere but it isn't clear why that is the case. In particular, I don't know if the 3.5.1 files above are sufficient, or whether we actually need something from 3.5.1.1 specifically.

**Comment by danburkert** [comment]
> Given that, and given that we seem to both want "cargo fetch" semantics anyway, if you're OK vendoring the binaries into the GitHub repo, I'd rather just skip to that.

Great.

The `protobuf` crate is special because it's only used for tests, it's never published to crates.io.  Given that, I think it's fine to leave it as is.  It could be switched to use a git submodule of the protobuf repo, but definitely not necessary to do as part of this.

> Or you want to have all platforms for which Google ships binary releases

I think this is probably the way to go, given that there's only a few more total, and I believe the current code tries to supports all of them.

The reason the `protobuf` crates dependency is slightly ahead is that the protobuf project didn't publish binaries for 3.5.1.1, so it wasn't possible to bump the protoc versions.  To be honest the `protoc` version really doesn't _have_ to be the latest release version; `prost` doesn't use any of the newer features.  On the other hand I like to keep the version in the `protobuf` crate up-to-date because it contains the upstream conformance tests which `prost` tests against, and they are actively improving those release to release.

**Comment by danburkert** [comment]
Brian Smith fixed this in #79, and it's released in 0.3.2.
